[aotinductor] Pass TorchIR to AOTInductor

angelayi opened the pull request:
Updates `_export.aot_compile` to pass a torch IR graph to inductor, allowing inductor to now run the pre_grad_passes, and reuse more of inductor's code. 
Also updates the API to only return the `so_path`, and not returning the exported program. The pytree call spec is now serialized and placed inside of the generated model code. When calling the model, because there is no c++ pytree implementation linked yet, we can access the call specs through `get_call_spec()`, and call pytree flatten/unflattenin python.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @avikchaudhuri @gmagogsfm @zhxchen17 @tugsbayasgalan

desertfire commented on benchmarks/dynamo/common.py:
```diff
@@ -1200,15 +1192,21 @@ def opt_export(_, example_inputs):
-    module, exported = AOTInductorModelCache.load(model, example_inputs)
+    module = AOTInductorModelCache.load(model, example_inputs)
+    call_spec = module.get_call_spec()
+    call_spec = (
```
Better to assign to two variables with explicit naming

desertfire commented on test/inductor/test_aot_inductor.py:
```diff
@@ -58,18 +59,20 @@ class AOTInductorModelRunner:
     def load(cls, so_path, example_inputs):
-        is_cpu = all(x.device.type == "cpu" for x in example_inputs)
+        is_cpu = all(
+            x.device.type == "cpu" for x in pytree.tree_flatten(example_inputs)[0]
```
Are you only checking the first input now?

angelayi replied:
pytree.tree_flatten returns a flattened list of example_inputs and the tree_spec for how it flattened the original list. So indexing [0] just gets the entire flattened list.

desertfire commented on test/inductor/test_aot_inductor.py:
```diff
@@ -81,33 +84,36 @@ def load(cls, so_path, example_inputs):
 
-        return optimized
+            call_spec = module.get_call_spec()
+            call_spec = (
```
Similar suggestion for the naming here

desertfire commented on torch/_inductor/codegen/aoti_runtime/interface.cpp:
```diff
@@ -129,6 +129,19 @@ AOTIRuntimeError AOTInductorModelContainerGetOutputName(
 
+AOTIRuntimeError AOTInductorModelContainerGetCallSpec(
+    AOTInductorModelContainerHandle container_handle,
+    std::string& in_spec,
```
These interface calls should only use C. Let's change it to `char *`.

angelayi committed:
[aotinductor] Pass TorchIR to AOTInductor (#110020)

Summary:
Updates `_export.aot_compile` to pass a torch IR graph to inductor, allowing inductor to now run the pre_grad_passes, and reuse more of inductor's code.
Also updates the API to only return the `so_path`, and not returning the exported program. The pytree call spec is now serialized and placed inside of the generated model code. When calling the model, because there is no c++ pytree implementation linked yet, we can access the call specs through `get_call_spec()`, and call pytree flatten/unflattenin python.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler avikchaudhuri gmagogsfm zhxchen17 tugsbayasgalan


Test Plan: Internal testing results: https://www.internalfb.com/diff/D49599792?dst_version_fbid=873731594363746&transaction_fbid=659025106213571

Reviewed By: frank-wei, desertfire

Differential Revision: D49599792

Pulled By: angelayi
